File: csharp/Snet/Rereader.cs
using System;
using System.IO;

namespace Snet
{
	internal class RereadData {
		public byte[] Data;
		public int Offset;
		public RereadData Next;
	}

	internal class Rereader {
		private RereadData _Head;
		private RereadData _Tail;
		private ulong _Count;

		public ulong Count {
			get { return _Count; }
		}

		public int Pull(byte[] buffer, int offset, int size) {
			if (_Head != null) {
				int headRemind = _Head.Data.Length - _Head.Offset;
				int count = headRemind < size ? headRemind : size;
				Buffer.BlockCopy (_Head.Data, _Head.Offset, buffer, offset, count);
				_Head.Offset += count;
				if (_Head.Offset >= _Head.Data.Length) {
					_Head = _Head.Next;
					if (_Head == null) {
						_Tail = null;
					}
				}
				_Count -= (ulong)count;
				return count;
			}
			return 0;
		}

		public bool Reread(Stream stream, int n) {
			byte[] b = new byte[n];
			try {
				for (int x = n; x >0; ) {
					x -= stream.Read(b, n - x, x);
					if (x == n)
						return false;
				}
			} catch {
				return false;
			}
			RereadData data = new RereadData ();
			data.Data = b;
			if (_Head == null) {
				_Head = data;
			} else {
				_Tail.Next = data;
			}
			_Tail = data;
			_Count += (ulong)n;
			return true;
		}
	}
}

File: go/rereader.go
package snet

import (
	"io"
)

type rereader struct {
	head *rereadData
	tail *rereadData
}

type rereadData struct {
	Data []byte
	next *rereadData
}

func (r *rereader) Pull(b []byte) (n int) {
	if r.head != nil {
		copy(b, r.head.Data)
		if len(r.head.Data) > len(b) {
			r.head.Data = r.head.Data[len(b):]
			n = len(b)
			return
		}
		n = len(r.head.Data)
		r.head = r.head.next
		if r.head == nil {
			r.tail = nil
		}
	}
	return
}

func (r *rereader) Reread(rd io.Reader, n int) bool {
	b := make([]byte, n)
	if _, err := io.ReadFull(rd, b); err != nil {
		return false
	}
	data := &rereadData{b, nil}
	if r.head == nil {
		r.head = data
	} else {
		r.tail.next = data
	}
	r.tail = data
	return true
}

func (r *rereader) Rollback() {
	if r.head == r.tail {
		r.head = nil
		r.tail = nil
		return
	}

	for item := r.head; item != nil; item = item.next {
		if item.next == r.tail {
			r.tail = item
			item.next = nil
			break
		}
	}
}
